### ISSUE TITLE
Make experiments reproducible with identical seeds and environments.

### ISSUE THREAD

**Issue by andrewargatkiny** [andrewargatkiny/dense-attention#39]
Currently, running experiments for many types of tasks several times in a row produces slightly different results (loss and accuracy curves). This happens despite exactly identical hyper-parameters and training parameters (including seed), packages and other software versions, same machine, same environment, etc. Affected categories of experiments include LRA and GLUE fine-tuning. Language model pre-training seems to be affected, too, but to a lesser degree.

This variability shouldn't be the case because the random seed is explicitly set at the start of training run for seemingly every random generator  (see e.g. the following snippet).

https://github.com/andrewargatkiny/dense-attention/blob/9242fe79d209ca4b7a40279d63f303fb34c65760/deepspeed_train.py#L642-L645

Seeds are also fixed during initialization of datasets (but maybe something somewhere was missed).

We need to trace the source of this randomness in the framework and eliminate it, so that experiments with identical seeds and all other parameters help equal yield identical results.
